app/models/secret_storage.rb

```diff
@@ -50,6 +50,43 @@ def self.keys
     end
   end
 
+  class HashicorpVault
+
+    VAULT_SECRET_BACKEND ='secret/'.freeze
+    # we don't really want other directories in here,
+    # and there may be other chars that we find we don't like
+    ENCODINGS = {"/": "%2F"}
+
+    def self.read(key)
+      result = Vault.logical.read(VAULT_SECRET_BACKEND  + self.encode(key))
+      result.data[:secret_data]
+    end
+
+    def self.write(key, data)
+      Vault.logical.write(VAULT_SECRET_BACKEND + self.encode(key), secret_data: data)
+    end
+
+    def self.delete(key)
+      Vault.logical.delete(VAULT_SECRET_BACKEND + key)
+    end
+
+    def self.keys()
+      Vault.logical.list(VAULT_SECRET_BACKEND).map! { |key| self.decode(key) }
+    end
+
+    private
+
+    def self.encode(string)
+      ENCODINGS.each { |k, v| string.gsub!(k.to_s, v.to_s) }
+      string
+    end
+
+    def self.decode(string)
```

Comment:
modifies the key -> need a bang ... but ideally don't ... `ENCODINGS.each { |k, v| string = string.gsub(v.to_s, k.to_s) }`